float32 direction   # [-1,0,1]
---